engine-src/Game/LambdaHack/Client/UI.hs

```diff
@@ -148,25 +149,34 @@ humanCommand = do
               modifySession $ \sess -> sess {sreportNull = True}
               -- Display base frame at the end.
               return []
-        LastRecord seqCurrent seqPrevious k <- getsSession slastRecord
-        let slastRecord
-              | k == 0 = LastRecord [] seqCurrent 0
-              | otherwise = LastRecord [] (seqCurrent ++ seqPrevious) (k - 1)
-        modifySession $ \sess -> sess {slastRecord}
         leader <- getLeaderUI
         b <- getsState $ getActorBody leader
         when (bhp b <= 0 && Just leader /= mOldLeader) $ displayMore ColorBW
           "If you move, the exertion will kill you. Consider asking for first aid instead."
         km <- promptGetKey ColorFull (EM.fromList [(propFont, over)]) False []
+        recording <- getsSession srecording
+        when recording $ do
+           -- If we're recording a in-game macro, we want to record each key.
+           macro <- getsSession smacroBuffer
+           modifySession $ \sess -> sess {smacroBuffer = km : macro}
         abortOrCmd <- do
           -- Look up the key.
           CCUI{coinput=InputContent{bcmdMap}} <- getsSession sccui
           case km `M.lookup` bcmdMap of
+            Just (_, _, RepeatLast) -> do
+              -- We can repeat every last action except 'repeat last action'
+              -- action, so we don't record that in last action's buffer.
+              modifySession $ \sess -> 
+                sess { swaitTimes = if swaitTimes sess > 0
+                                    then - swaitTimes sess
+                                    else 0 }
+              cmdHumanSem RepeatLast
             Just (_, _, cmd) -> do
-              modifySession $ \sess -> sess
-                {swaitTimes = if swaitTimes sess > 0
-                              then - swaitTimes sess
-                              else 0}
+              modifySession $ \sess ->
+                sess { swaitTimes = if swaitTimes sess > 0
+                                    then - swaitTimes sess
+                                    else 0
+                     , slastAction = Just km } -- Record user's last action.
```

Comment:
Nitpicking now: `slastAction = if cmd == RepeatLast then slastAction sess else Just km`

Reply:
To be clear: then we can deduplicate the surrounding code.